In [1]:
import json  
import pandas as pd  

# Inicializa una lista vacía para almacenar los datos extraídos del archivo JSON
rows = []

# Abre el archivo JSON en modo de lectura ('r'). Especificar la codificación ('utf-8')
with open('DataSet/output_steam_games.json', 'r', encoding='utf-8') as f:
    # Iterar sobre cada línea del archivo
    for line in f.readlines():
        # Convertir cada línea a un diccionario de Python y agregarlo a la lista "rows"
        rows.append(json.loads(line))

# Crear un DataFrame a partir de la lista de diccionarios 'rows'
df_sg = pd.DataFrame(rows)


In [2]:
df_sg

publisher                                 genres  \
0                   NaN                                    NaN   
1                   NaN                                    NaN   
2                   NaN                                    NaN   
3                   NaN                                    NaN   
4                   NaN                                    NaN   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444              NaN                                    NaN   

                        app_name                     title  \
0                            NaN                       NaN   
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                       NaN   

                                                      url release_date  \
0                                                     NaN          NaN   
1                                                     NaN          NaN   
2                                                     NaN          NaN   
3                                                     NaN          NaN   
4                                                     NaN          NaN   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                     tags  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440  http://steamcommunity.com/app/773640/reviews/?...   
120441  http://steamcommunity.com/app/733530/reviews/?...   
120442  http://steamcommunity.com/app/610660/reviews/?...   
120443  http://steamcommunity.com/app/658870/reviews/?...   
120444  http://steamcommunity.com/app/681550/reviews/?...   

                                                    specs price early_access  \
0                                        

In [3]:
df_sg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120445 entries, 0 to 120444
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher     24083 non-null  object
 1   genres        28852 non-null  object
 2   app_name      32133 non-null  object
 3   title         30085 non-null  object
 4   url           32135 non-null  object
 5   release_date  30068 non-null  object
 6   tags          31972 non-null  object
 7   reviews_url   32133 non-null  object
 8   specs         31465 non-null  object
 9   price         30758 non-null  object
 10  early_access  32135 non-null  object
 11  id            32133 non-null  object
 12  developer     28836 non-null  object
dtypes: object(13)
memory usage: 11.9+ MB


In [4]:
# Contar filas con nulos en todas las columnas
rows_with_all_nulls = df_sg[df_sg.isnull().all(axis=1)]

# Mostrar el número de filas con nulos en todas las columnas
print("Número de filas con nulos en todas las columnas:", len(rows_with_all_nulls))


Número de filas con nulos en todas las columnas: 88310


In [5]:
# Eliminar las filas con nulos en todas las columnas simultaneamente 
df_sg = df_sg.dropna(how='all')

# Verificar que las filas se han eliminado correctamente
print("Número de filas después de eliminar nulos en todas las columnas:", len(df_sg))

Número de filas después de eliminar nulos en todas las columnas: 32135


In [6]:
df_sg.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32135 entries, 88310 to 120444
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher     24083 non-null  object
 1   genres        28852 non-null  object
 2   app_name      32133 non-null  object
 3   title         30085 non-null  object
 4   url           32135 non-null  object
 5   release_date  30068 non-null  object
 6   tags          31972 non-null  object
 7   reviews_url   32133 non-null  object
 8   specs         31465 non-null  object
 9   price         30758 non-null  object
 10  early_access  32135 non-null  object
 11  id            32133 non-null  object
 12  developer     28836 non-null  object
dtypes: object(13)
memory usage: 3.4+ MB


In [7]:
# Estadísticas descirptivas del DataFrame
df_sg.describe()

publisher    genres    app_name       title  \
count      24083     28852       32133       30085   
unique      8239       883       32094       30054   
top      Ubisoft  [Action]  Soundtrack  Soundtrack   
freq         385      1880           3           3   

                                                      url release_date  \
count                                               32135        30068   
unique                                              32135         3582   
top     http://store.steampowered.com/app/761140/Lost_...   2012-10-16   
freq                                                    1          100   

                        tags  \
count                  31972   
unique                 15395   
top     [Casual, Simulation]   
freq                    1292   

                                              reviews_url            specs  \
count                                               32133            31465   
unique                                              32132             4649   
top     http://steamcommunity.com/app/612880/reviews/?...  [Single-player]   
freq                                                    2             2794   

           price early_access      id                developer  
count   30758.00        32135   32133                    28836  
unique    162.00            2   32132                    10992  
top         4.99        False  612880  Ubisoft - San Francisco  
freq     4278.00        30188       2                     1259

## Columna genres

In [8]:
# Imprimir filas de la columna 'genres'
print(df_sg['genres'].head())

88310        [Action, Casual, Indie, Simulation, Strategy]
88311                 [Free to Play, Indie, RPG, Strategy]
88312    [Casual, Free to Play, Indie, Simulation, Sports]
88313                          [Action, Adventure, Casual]
88314                                                  NaN
Name: genres, dtype: object


In [9]:
# Imprimir la cantidad de valores nulos en la columna 'genres'
print("Número de valores nulos en 'genres':", df_sg['genres'].isnull().sum())

# Imprimir la cantidad de valores no nulos en 'genres'
print("Número de valores no nulos en 'genres':", df_sg['genres'].notnull().sum())

Número de valores nulos en 'genres': 3283
Número de valores no nulos en 'genres': 28852


## Columna tags

In [10]:
# Imprimir filas de la columna 'tags'
print(df_sg['tags'].head())

88310        [Strategy, Action, Indie, Casual, Simulation]
88311    [Free to Play, Strategy, Indie, RPG, Card Game...
88312    [Free to Play, Simulation, Sports, Casual, Ind...
88313                          [Action, Adventure, Casual]
88314                      [Action, Indie, Casual, Sports]
Name: tags, dtype: object


In [11]:
# Imprimir la cantidad de valores nulos en la columna 'tags'
print("Número de valores nulos en 'tags':", df_sg['tags'].isnull().sum())

# Imprimir la cantidad de valores no nulos en 'tags'
print("Número de valores no nulos en 'tags':", df_sg['tags'].notnull().sum())

Número de valores nulos en 'tags': 163
Número de valores no nulos en 'tags': 31972


### Rellenar filas vacías de la columna genres con los valores de la columna tags

Al observar la similitud entre la columna genres y tags, para evitar eliminar nulos y perder información valiosa para nuestro análisis, optamos por rellenar los valores faltantes de la columna genres, con los valores de la columna tags.

In [12]:
# Convertir las listas en cadenas en ambas columnas
df_sg['genres_str'] = df_sg['genres'].astype(str)
df_sg['tags_str'] = df_sg['tags'].astype(str)

# Rellenar las filas vacías en 'genres' con los valores de 'tags'
df_sg['genres_filled'] = df_sg['genres_str'].fillna(df_sg['tags_str'])

# Eliminar columnas auxiliares
df_sg.drop(['genres_str', 'tags_str'], axis=1, inplace=True)

# Mostrar información después del proceso
print("Número de valores nulos en 'genres_filled':", df_sg['genres_filled'].isnull().sum())

Número de valores nulos en 'genres_filled': 0


C:\Users\nicol\AppData\Local\Temp\ipykernel_3088\43977612.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sg['genres_str'] = df_sg['genres'].astype(str)
C:\Users\nicol\AppData\Local\Temp\ipykernel_3088\43977612.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sg['tags_str'] = df_sg['tags'].astype(str)
C:\Users\nicol\AppData\Local\Temp\ipykernel_3088\43977612.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] 

In [13]:
# Imprimir la cantidad de valores nulos en la columna 'genres_filled'
print("Número de valores nulos en 'genres_filled':", df_sg['genres_filled'].isnull().sum())

# Imprimir la cantidad de valores no nulos en 'genres_filled'
print("Número de valores no nulos en 'genres_filled':", df_sg['genres_filled'].notnull().sum())

Número de valores nulos en 'genres_filled': 0
Número de valores no nulos en 'genres_filled': 32135


In [14]:
# Eliminar las columnas 'genres' y 'tags'
df_sg.drop(['genres', 'tags'], axis=1, inplace=True)

C:\Users\nicol\AppData\Local\Temp\ipykernel_3088\479775121.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sg.drop(['genres', 'tags'], axis=1, inplace=True)


In [15]:
# Renombrar la columna
df_sg.rename(columns={'genres_filled': 'genres'}, inplace=True)

C:\Users\nicol\AppData\Local\Temp\ipykernel_3088\3000696023.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sg.rename(columns={'genres_filled': 'genres'}, inplace=True)


In [16]:
# Reemplazar los corchetes con cadenas vacías y las comillas vacías('') en la columna 'genres'
df_sg['genres'] = df_sg['genres'].str.replace(r'\[|\]|\'' , '', regex=True)

# Imprimir las primeras filas para verificar los cambios
print(df_sg['genres'].head())

88310        Action, Casual, Indie, Simulation, Strategy
88311                 Free to Play, Indie, RPG, Strategy
88312    Casual, Free to Play, Indie, Simulation, Sports
88313                          Action, Adventure, Casual
88314                                                nan
Name: genres, dtype: object


C:\Users\nicol\AppData\Local\Temp\ipykernel_3088\3313765434.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sg['genres'] = df_sg['genres'].str.replace(r'\[|\]|\'' , '', regex=True)


In [17]:
# Imprimir solo la columna genres_filled
print(df_sg['genres'])


88310         Action, Casual, Indie, Simulation, Strategy
88311                  Free to Play, Indie, RPG, Strategy
88312     Casual, Free to Play, Indie, Simulation, Sports
88313                           Action, Adventure, Casual
88314                                                 nan
                               ...                       
120440                Casual, Indie, Simulation, Strategy
120441                            Casual, Indie, Strategy
120442                          Indie, Racing, Simulation
120443                                      Casual, Indie
120444                                                nan
Name: genres, Length: 32135, dtype: object


## Columna app_name

In [18]:
# Imprimir filas de la columna 'app_name'
print(df_sg['app_name'].head())

88310        Lost Summoner Kitty
88311                  Ironbound
88312    Real Pool 3D - Poolians
88313                    弹炸人2222
88314              Log Challenge
Name: app_name, dtype: object


In [19]:
# Imprimir la cantidad de valores nulos en la columna 'app_name'
print("Número de valores nulos en 'app_name':", df_sg['app_name'].isnull().sum())

# Imprimir la cantidad de valores no nulos en 'title'
print("Número de valores no nulos en 'app_name':", df_sg['app_name'].notnull().sum())

Número de valores nulos en 'app_name': 2
Número de valores no nulos en 'app_name': 32133


In [20]:
# Contar el número de valores únicos
unique_app_name_count = df_sg['app_name'].nunique()

# Imprime el número de valores únicos encontrados en la columna 'app_name'.
print(f"Número de valores únicos en 'app_name': {unique_app_name_count}")


Número de valores únicos en 'app_name': 32094


## Columna title

In [21]:
# Imprimir filas de la columna 'title'
print(df_sg['title'].head())

88310        Lost Summoner Kitty
88311                  Ironbound
88312    Real Pool 3D - Poolians
88313                    弹炸人2222
88314                        NaN
Name: title, dtype: object


In [22]:
# Imprimir la cantidad de valores nulos en la columna 'title'
print("Número de valores nulos en 'title':", df_sg['title'].isnull().sum())

# Imprimir la cantidad de valores no nulos en 'title'
print("Número de valores no nulos en 'title':", df_sg['title'].notnull().sum())

Número de valores nulos en 'title': 2050
Número de valores no nulos en 'title': 30085


In [23]:
# Calcula el número de títulos únicos presentes en la columna 'title'
unique_title_count = df_sg['title'].nunique()

# Imprime el número de títulos únicos en 'title'
print(f"Número de valores únicos en 'title': {unique_title_count}")


Número de valores únicos en 'title': 30054


### Rellenar filas vacías de la columna title con los valores de la columna app_name

Al observar la similitud entre la columna app_name y title, para evitar eliminar filas con nulos y perder información valiosa para nuestro análisis, optamos por rellenar los valores faltantes de la columna title, con los valores de la columna app_name.

In [24]:
# Rellenar los valores nulos en 'title' con los valores de 'app_name'
df_sg['title'].fillna(df_sg['app_name'], inplace=True)


C:\Users\nicol\AppData\Local\Temp\ipykernel_3088\3475956799.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sg['title'].fillna(df_sg['app_name'], inplace=True)


In [25]:
# Imprimir la cantidad de valores nulos en la columna 'title'
print("Número de valores nulos en 'title':", df_sg['title'].isnull().sum())

# Imprimir la cantidad de valores no nulos en 'title'
print("Número de valores no nulos en 'title':", df_sg['title'].notnull().sum())

Número de valores nulos en 'title': 2
Número de valores no nulos en 'title': 32133


In [26]:
# Eliminar la columna 'app_name'
df_sg.drop(columns=['app_name'], inplace=True)

C:\Users\nicol\AppData\Local\Temp\ipykernel_3088\208677527.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sg.drop(columns=['app_name'], inplace=True)


Revisamos si la nueva columna title tiene valores duplicados

In [27]:
duplicates = df_sg[df_sg.duplicated('title', keep=False)]
print(duplicates[['title']])

                                                 title
88384                                              NaN
89378   Batman: Arkham City - Game of the Year Edition
89818                     The Dream Machine: Chapter 4
89819                     The Dream Machine: Chapter 4
89975                                      Dark Matter
...                                                ...
118857                             AirMech® Soundtrack
119271  Batman: Arkham City - Game of the Year Edition
119713                                       Alter Ego
119852                                        Altitude
120307                               MORE SWEATER? OK!

[77 rows x 1 columns]


Los valores duplicados no aportan al proyecto. Además, no se justifican en esta columna, pues no se trata, por ejemplo, de versiones distintas del mismo juego. Asumimos que es información redundante y no aporta parsimonia al proyecto. Procedemos a eliminarlos.

In [28]:
# Eliminar duplicados
df_sg = df_sg.drop_duplicates(subset='title', keep='first')

In [29]:
# Revisar nulos
df_sg['title'].isnull().sum()

1

In [30]:
# Eliminar filas con valores nulos en la columna 'title'
df_sg = df_sg.dropna(subset=['title'])

## Columna Publisher

In [31]:
# Mostrar la distribución general de los valores no nulos en la columna 'publisher'
publisher_distribution = df_sg['publisher'].value_counts()
print("Distribución de 'publisher':")
print(publisher_distribution)

Distribución de 'publisher':
publisher
Ubisoft                    385
Dovetail Games - Trains    360
Degica                     286
Paradox Interactive        271
SEGA                       250
                          ... 
Atomic Wolf                  1
Rockywood Studios            1
Crabellum Games              1
Auburn Sounds                1
Bidoniera Games              1
Name: count, Length: 8228, dtype: int64


In [32]:
# Obtener el número total de valores únicos en la columna 'publisher'
total_unique_publishers = df_sg['publisher'].nunique()
print("\nNúmero total de editores únicos:", total_unique_publishers)


Número total de editores únicos: 8228


In [33]:
# Contar nulos y no nulos en la columna 'publisher'
null_count = df_sg['publisher'].isnull().sum()
not_null_count = df_sg['publisher'].notnull().sum()

print("Número de valores nulos en 'publisher':", null_count)
print("Número de valores no nulos en 'publisher':", not_null_count)

Número de valores nulos en 'publisher': 8042
Número de valores no nulos en 'publisher': 24053


## Columna developer

In [34]:
# Mostrar la distribución general de los valores no nulos en la columna 'developer'
developer_counts = df_sg['developer'].value_counts()
print("Frecuencia de cada desarrollador:", developer_counts)

Frecuencia de cada desarrollador: developer
Ubisoft - San Francisco                     1259
SmiteWorks USA, LLC                          813
Dovetail Games                               253
KOEI TECMO GAMES CO., LTD.                   232
Paradox Development Studio                   156
                                            ... 
Class of  2016-2017                            1
Pirotexnik                                     1
ElVery333,Raquel Covelo Suarez,ATopeCode       1
MNB                                            1
Bidoniera Games                                1
Name: count, Length: 10979, dtype: int64


In [35]:
# Contar nulos y no nulos en la columna 'publisher'
print("Número de valores nulos en 'developer':", df_sg['developer'].isnull().sum())
print("Número de valores no nulos en 'developer':", df_sg['developer'].notnull().sum())

Número de valores nulos en 'developer': 3290
Número de valores no nulos en 'developer': 28805


In [36]:
# Obtener el número total de valores únicos en la columna 'developer'
total_unique_publishers = df_sg['developer'].nunique()
print("\nNúmero total de desarrolladores únicos:", total_unique_publishers)


Número total de desarrolladores únicos: 10979


### Rellenar filas vacías de la columna developer con los valores de la columna publisher

La columna "Publisher" hace alusion a la entidad responsable de distribuir y comercializar el juego. Si bien, no son los mismos valores en todas las filas de "developer" y "publisher", en muchos casos coinciden. Es preferible tener información al menos de quien comercializa el juego, antes que ninguna información en la columna "developer". De esta forma, logramos salvar filas que cuentan con información imortante para el presente proyecto.

In [37]:
# Rellenar las filas vacías en 'developer' con los valores de 'publisher'
df_sg['developer_filled'] = df_sg['developer'].fillna(df_sg['publisher'])

# Mostrar información después del proceso
print("Número de valores nulos en 'developer_filled':", df_sg['developer_filled'].isnull().sum()) 

# Eliminar la columna 'publisher'
df_sg = df_sg.drop('publisher', axis=1)

# Eliminar la columna 'developer'
df_sg.drop(['developer'], axis=1, inplace=True)

Número de valores nulos en 'developer_filled': 3225


In [38]:
# Renombrar la columna
df_sg.rename(columns={'developer_filled': 'developer'}, inplace=True)

Revisamos ahora los valores nulos de la nueva columna que hemos creado.

In [39]:
# Revisar nulos
null_developers = df_sg['developer'].isnull().sum()
print(f"Número de valores nulos en la columna 'developer': {null_developers}")

Número de valores nulos en la columna 'developer': 3225


Exploramos la naturaleza de los nulos de la columna developer. Para ello tomamos una muestra de los valores nulos de la columna  y la comparamos con las mismas filas en las columnas title y genres

In [40]:
# Filtrar filas con valores nulos en 'developer'
null_developer_rows = df_sg[df_sg['developer'].isnull()]

# Mostrar algunas filas específicas
print(null_developer_rows[['title', 'genres']])

                                             title genres
88314                                Log Challenge    nan
88321                         Icarus Six Sixty Six    nan
88329                                After Life VR    nan
88330                                   Kitty Hawk    nan
88332                                   Mortars VR    nan
...                                            ...    ...
120428   Robotpencil Presents: Exercise: Brushwork    nan
120429  Robotpencil Presents: Creative Composition    nan
120430                            The Gamble House    nan
120431    Kalen Chock Presents: 2017 Free Tutorial    nan
120444                                 Maze Run VR    nan

[3225 rows x 2 columns]


In [41]:
# Contar la cantidad de filas con 'nan' y con géneros específicos
nan_genre_count = len(df_sg[df_sg['genres'] == 'nan'])
specific_genre_count = len(df_sg[df_sg['genres'] != 'nan'])

print("Número de filas con 'nan' en 'genres':", nan_genre_count)
print("Número de filas con géneros específicos:", specific_genre_count)

Número de filas con 'nan' en 'genres': 3275
Número de filas con géneros específicos: 28820


Podemos advertir que las filas con nulos en la columna developer también lo son en genres. Estos valores parecen estar relacionados,  por lo que eliminarlos podría mejorar la consistencia del conjunto de datos.

In [42]:
# Eliminar los nulos de la columna developer
df_sg = df_sg.dropna(subset=['developer'])

## Columna url

In [43]:
# Imprimir columna "url"
print(df_sg['url'])

88310     http://store.steampowered.com/app/761140/Lost_...
88311     http://store.steampowered.com/app/643980/Ironb...
88312     http://store.steampowered.com/app/670290/Real_...
88313        http://store.steampowered.com/app/767400/2222/
88315     http://store.steampowered.com/app/772540/Battl...
                                ...                        
120439    http://store.steampowered.com/app/745400/Kebab...
120440    http://store.steampowered.com/app/773640/Colon...
120441    http://store.steampowered.com/app/733530/LOGis...
120442    http://store.steampowered.com/app/610660/Russi...
120443    http://store.steampowered.com/app/658870/EXIT_...
Name: url, Length: 28870, dtype: object


La información que nos entrega la columna "url" es irrelevante a efectos del presente proyecto.

In [44]:
# Eliminar columna "url"
df_sg.drop(columns=['url'], inplace=True)

## Columna release_date

In [45]:
# Imprimir filas de la columna 'release_date'
print(df_sg['release_date'].head())

88310    2018-01-04
88311    2018-01-04
88312    2017-07-24
88313    2017-12-07
88315    2018-01-04
Name: release_date, dtype: object


In [46]:
# Imprimir la cantidad de valores nulos en la columna 'release_date'
print("Número de valores nulos en 'release_date':", df_sg['release_date'].isnull().sum())

# Imprimir la cantidad de valores no nulos en 'release_date'
print("Número de valores no nulos en 'release_date':", df_sg['release_date'].notnull().sum())

Número de valores nulos en 'release_date': 18
Número de valores no nulos en 'release_date': 28852


A efectos de las funciones a desarrollar, así como el modelo de recomendación, sólo necesitamos el año de la columna release_date.

In [47]:
# Convertir la columna 'release_date' a formato de fecha
df_sg['release_date'] = pd.to_datetime(df_sg['release_date'], errors='coerce')

# Crear una nueva columna 'release_year' que contiene el año
df_sg['release_year'] = df_sg['release_date'].dt.year

# Imprimir la nueva columna 'release_year'
print(df_sg['release_year'])


88310     2018.0
88311     2018.0
88312     2017.0
88313     2017.0
88315     2018.0
           ...  
120439    2018.0
120440    2018.0
120441    2018.0
120442    2018.0
120443    2017.0
Name: release_year, Length: 28870, dtype: float64


In [48]:
# Imprimir la cantidad de valores nulos en la columna 'release_year'
print("Número de valores nulos en 'release_year':", df_sg['release_year'].isnull().sum())

# Imprimir la cantidad de valores no nulos en 'release_year'
print("Número de valores no nulos en 'release_year':", df_sg['release_year'].notnull().sum())

Número de valores nulos en 'release_year': 302
Número de valores no nulos en 'release_year': 28568


Los valores nulos en 'release_year' representan una pequeña fracción del total (302 de 28870, es aproximadamente el 1%). Por ello, decidimos imputar estos valores faltantes con la media pues, al ser una proporción tan baja, no distorsiona significativamente la distribución general de los datos. Así podemos salvar estas filas que contienen información relevante para nuestro propósito.

In [49]:
# Imputar la media de 'release_year' a los nulos
mean_release_year = df_sg['release_year'].mean()
df_sg['release_year'].fillna(mean_release_year, inplace=True)


In [50]:
# Eliminar la columna 'release_date'
df_sg.drop(columns=['release_date'], inplace=True)


## Columna reviews_url

In [51]:
# Imprimir la columna 'reviews_url'
print(df_sg['reviews_url'])


88310     http://steamcommunity.com/app/761140/reviews/?...
88311     http://steamcommunity.com/app/643980/reviews/?...
88312     http://steamcommunity.com/app/670290/reviews/?...
88313     http://steamcommunity.com/app/767400/reviews/?...
88315     http://steamcommunity.com/app/772540/reviews/?...
                                ...                        
120439    http://steamcommunity.com/app/745400/reviews/?...
120440    http://steamcommunity.com/app/773640/reviews/?...
120441    http://steamcommunity.com/app/733530/reviews/?...
120442    http://steamcommunity.com/app/610660/reviews/?...
120443    http://steamcommunity.com/app/658870/reviews/?...
Name: reviews_url, Length: 28870, dtype: object


El contenido de la columna "reviews_url" no es relevante para el presente proyecto, por lo que podemos prescindir de ella.

In [52]:
# Eliminar la columna 'reviews_url'
df_sg.drop('reviews_url', axis=1, inplace=True)


## Columna specs

In [53]:
# Imprimir la columna 'specs'
print(df_sg['specs'])

88310                                       [Single-player]
88311     [Single-player, Multi-player, Online Multi-Pla...
88312     [Single-player, Multi-player, Online Multi-Pla...
88313                                       [Single-player]
88315                   [Single-player, Steam Achievements]
                                ...                        
120439     [Single-player, Steam Achievements, Steam Cloud]
120440                  [Single-player, Steam Achievements]
120441    [Single-player, Steam Achievements, Steam Clou...
120442    [Single-player, Steam Achievements, Steam Trad...
120443     [Single-player, Steam Achievements, Steam Cloud]
Name: specs, Length: 28870, dtype: object


El contenido de esta columna no es útil para nuestro proyecto.

In [54]:
# Eliminar la columna "specs"
df_sg = df_sg.drop('specs', axis=1)


## Columna price

In [55]:
# Imprimir filas de la columna 'price'
print(df_sg['price'].head())

88310            4.99
88311    Free To Play
88312    Free to Play
88313            0.99
88315            3.99
Name: price, dtype: object


In [56]:
# Número de valores nulos en 'price'
null_count_price = df_sg['price'].isnull().sum()

# Número de valores no nulos en 'price'
non_null_count_price = df_sg['price'].notnull().sum()

# Imprimir los resultados
print(f"Número de valores nulos en 'price': {null_count_price}")
print(f"Número de valores no nulos en 'price': {non_null_count_price}")

Número de valores nulos en 'price': 1235
Número de valores no nulos en 'price': 27635


In [57]:
# Valores únicos en la columna 'price'
unique_prices = df_sg['price'].unique()

# Imprimir los resultados
print("Valores únicos en la columna 'price':", unique_prices)

Valores únicos en la columna 'price': [4.99 'Free To Play' 'Free to Play' 0.99 3.99 9.99 18.99 29.99 nan 10.99
 2.99 1.59 14.99 1.99 59.99 8.99 6.99 7.99 39.99 'Free' 19.99 7.49 12.99
 5.99 2.49 15.99 1.25 24.99 17.99 61.99 3.49 11.99 13.99 'Free Demo'
 'Play for Free!' 34.99 74.76 1.49 32.99 99.99 14.95 69.99 16.99 79.99
 49.99 5.0 13.98 29.96 119.99 109.99 149.99 771.71 'Install Now' 21.99
 89.99 'Play WARMACHINE: Tactics Demo' 0.98 139.92 4.29 'Free Mod' 54.99
 64.99 74.99 'Install Theme' 0.89 'Third-party' 0.5 'Play Now' 299.99 1.29
 44.99 3.0 15.0 5.49 23.99 49.0 10.93 1.39 'Free HITMAN™ Holiday Pack'
 36.99 4.49 2.0 4.0 234.99 1.95 1.5 199.0 189.0 6.66 27.99 129.99 179.0
 26.99 399.99 31.99 399.0 20.0 40.0 3.33 22.99 320.0 38.85 71.7 995.0
 27.49 3.39 6.0 19.95 20.99 499.99 199.99 16.06 4.68 131.4 44.98 202.76
 2.3 0.95 172.24 249.99 2.97 10.96 10.0 30.0 2.66 6.48 1.0 11.15
 'Play the Demo' 99.0 87.94 0.49 9.98 9.95 7.0 12.89 6.49 1.87 42.99 41.99
 289.99 'Free to Try' 23.96 5.65

In [58]:
# Filtrar solo los valores no numéricos o que no son números flotantes
non_float_prices = df_sg['price'].loc[~df_sg['price'].astype(str).str.match(r'^\d+\.\d+$')].unique()

# Imprimir los resultados
print("Valores únicos que no son números con decimales en la columna 'price':", non_float_prices)

Valores únicos que no son números con decimales en la columna 'price': ['Free To Play' 'Free to Play' nan 'Free' 'Free Demo' 'Play for Free!'
 'Install Now' 'Play WARMACHINE: Tactics Demo' 'Free Mod' 'Install Theme'
 'Third-party' 'Play Now' 'Free HITMAN™ Holiday Pack' 'Play the Demo'
 'Free to Try' 'Free to Use']


In [59]:
# Lista de valores que deben convertirse a 0.00
non_float_values = ['Free To Play', 'Free to Play', 'Free', 'Free Demo', 'Play for Free!',
                    'Install Now', 'Play WARMACHINE: Tactics Demo', 'Free Mod', 'Install Theme',
                    'Third-party', 'Play Now', 'Free HITMAN™ Holiday Pack', 'Play the Demo',
                    'Starting at $499.00', 'Starting at $449.00', 'Free to Try', 'Free Movie',
                    'Free to Use']

# Reemplazar los valores específicos con 0.00 y mantener los NaN
df_sg['price'] = df_sg['price'].replace(non_float_values, 0.00)

In [60]:
# Convertir la columna 'price' a tipo de dato float
df_sg['price'] = pd.to_numeric(df_sg['price'], errors='coerce')

# Imprimir la columna 'price'
print(df_sg['price'])

88310     4.99
88311     0.00
88312     0.00
88313     0.99
88315     3.99
          ... 
120439    1.99
120440    1.99
120441    4.99
120442    1.99
120443    4.99
Name: price, Length: 28870, dtype: float64


In [61]:
# Número de valores nulos en 'price'
null_count_price = df_sg['price'].isnull().sum()

# Número de valores no nulos en 'price'
non_null_count_price = df_sg['price'].notnull().sum()

# Imprimir los resultados
print(f"Número de valores nulos en 'price': {null_count_price}")
print(f"Número de valores no nulos en 'price': {non_null_count_price}")

Número de valores nulos en 'price': 1235
Número de valores no nulos en 'price': 27635


Revisamos nuevamente los valores nulos

In [62]:
# Sumar nulos de la columna "price"
nulos_id = df_sg['price'].isnull().sum()
print(f"Número de valores nulos en la columna 'price': {nulos_id}")

Número de valores nulos en la columna 'price': 1235


In [63]:
# Calcular la proporción de valores nulos en 'price'
proporcion_nulos_price = df_sg['price'].isnull().sum() / len(df_sg)

# Imprime la proporción de valores nulos en 'price'
print(f"Proporción de valores nulos en 'price': {proporcion_nulos_price:.2%}")

Proporción de valores nulos en 'price': 4.28%


Dado que la proporción de valores nulos en la columna 'price' es baja (4.28%), eliminar estos registros nulos es razonable.

In [64]:
# Eliminar filas con valores nulos en la columna 'price'
df_sg = df_sg.dropna(subset=['price'])

## Columna early_access

In [65]:
# Imprimir filas de la columna 'early_access'
print(df_sg['early_access'].head())

88310    False
88311    False
88312    False
88313    False
88315    False
Name: early_access, dtype: object


In [66]:
# Valores nulos y no nulos en la columna 'early_access'
print("Número de valores nulos en 'early_access':", df_sg['early_access'].isnull().sum())
print("Número de valores no nulos en 'early_access':", df_sg['early_access'].count())

Número de valores nulos en 'early_access': 0
Número de valores no nulos en 'early_access': 27635


In [67]:
# Mostrar los valores únicos de la columna early_access
unique_values = df_sg['early_access'].unique()

# Imprimir los valores únicos
print(unique_values)


[False True]


In [68]:
# Convertir la columna 'early_access' a booleano
df_sg['early_access'] = df_sg['early_access'].astype(bool)

# Verificar el cambio
print(df_sg['early_access'].dtype)


bool


## Columna id

In [69]:
# Imprimir filas de la columna 'id'
print(df_sg['id'].head())

88310    761140
88311    643980
88312    670290
88313    767400
88315    772540
Name: id, dtype: object


In [70]:
# Nulos y no nulos en la columna 'id'
print("Nulos en 'id':", df_sg['id'].isnull().sum())
print("No nulos en 'id':", df_sg['id'].notnull().sum())

Nulos en 'id': 0
No nulos en 'id': 27635


In [71]:
# Convertir la columna 'id' a tipo entero
df_sg['id'] = pd.to_numeric(df_sg['id'], errors='coerce', downcast='integer')

Revisión del resultado de las transformaciones realizadas sobre el DataSet original output_steam_games.json

In [72]:
df_sg

title  price  early_access      id  \
88310        Lost Summoner Kitty   4.99         False  761140   
88311                  Ironbound   0.00         False  643980   
88312    Real Pool 3D - Poolians   0.00         False  670290   
88313                    弹炸人2222   0.99         False  767400   
88315      Battle Royale Trainer   3.99         False  772540   
...                          ...    ...           ...     ...   
120439              Kebab it Up!   1.99         False  745400   
120440            Colony On Mars   1.99         False  773640   
120441  LOGistICAL: South Africa   4.99         False  733530   
120442             Russian Roads   1.99         False  610660   
120443       EXIT 2 - Directions   4.99         False  658870   

                                                 genres  \
88310       Action, Casual, Indie, Simulation, Strategy   
88311                Free to Play, Indie, RPG, Strategy   
88312   Casual, Free to Play, Indie, Simulation, Sports   
88313                         Action, Adventure, Casual   
88315                     Action, Adventure, Simulation   
...                                                 ...   
120439                 Action, Adventure, Casual, Indie   
120440              Casual, Indie, Simulation, Strategy   
120441                          Casual, Indie, Strategy   
120442                        Indie, Racing, Simulation   
120443                                    Casual, Indie   

                       developer  release_year  
88310                  Kotoshiro        2018.0  
88311           Secret Level SRL        2018.0  
88312               Poolians.com        2017.0  
88313                       彼岸领域        2017.0  
88315        Trickjump Games Ltd        2018.0  
...                          ...           ...  
120439           Bidoniera Games        2018.0  
120440        Nikita "Ghost_RUS"        2018.0  
120441                    Sacada        2018.0  
120442  Laush Dmitriy Sergeevich        2018.0  
120443             xropi,stev3ns        2017.0  

[27635 rows x 7 columns]

In [73]:
df_sg.info()

<class 'pandas.core.frame.DataFrame'>
Index: 27635 entries, 88310 to 120443
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   title         27635 non-null  object 
 1   price         27635 non-null  float64
 2   early_access  27635 non-null  bool   
 3   id            27635 non-null  int32  
 4   genres        27635 non-null  object 
 5   developer     27635 non-null  object 
 6   release_year  27635 non-null  float64
dtypes: bool(1), float64(2), int32(1), object(3)
memory usage: 1.4+ MB


In [74]:
# Reiniciar el índice
df_sg = df_sg.reset_index(drop=True)

In [75]:
# Rutas de los archivos CSV y Parquet
# csv_path = 'DataSet/clean_df_sg.csv'
parquet_path = 'DataSet/clean_df_sg.parquet'

# Guardar el DataFrame como un archivo CSV
# df_sg.to_csv(csv_path, index=False)


# Guardar el DataFrame como un archivo Parquet
df_sg.to_parquet(parquet_path, index=False)
print(f'El archivo Parquet {parquet_path} ha sido creado.')

El archivo Parquet DataSet/clean_df_sg.parquet ha sido creado.
